py/farm_ng/core/uri.py

```diff
@@ -13,39 +17,26 @@ def __repr__(self) -> str:
         return self.string()
 
     @property
-    def proto(self) -> event_pb2.Uri:
+    def proto(self) -> uri_pb2.Uri:
         return self._uri
 
     @property
     def scheme(self) -> str:
-        scheme_value = event_pb2.UriSchemeType.DESCRIPTOR.values[self._uri.scheme]
+        scheme_value = uri_pb2.UriSchemeType.DESCRIPTOR.values[self._uri.scheme]
         return scheme_value.name.lower()
 
     @classmethod
     def empty(cls) -> "Uri":
-        return Uri(event_pb2.Uri())
+        return Uri(uri_pb2.Uri())
 
     def string(self) -> str:
         return (
             f"{self.scheme}://{self._uri.authority}/{self._uri.path}?{self._uri.query}"
         )
 
     @classmethod
-    def from_string(self, string: str) -> "Uri":
-        scheme_name, remainder = string.split("://")
-        remainder, query = remainder.split("?")
-        authority, path = remainder.split("/")
-        return Uri.from_strings(scheme_name, authority, path, query)
-
-    @classmethod
-    def from_strings(
-        self, scheme_name: str, authority: str, path: str, query: str
-    ) -> "Uri":
-        scheme_value = event_pb2.UriSchemeType.DESCRIPTOR.values_by_name[
-            scheme_name.upper()
-        ]
-        return Uri(
-            event_pb2.Uri(
-                scheme=scheme_value.number, authority=authority, path=path, query=query
-            )
+    def from_message(cls, message: Any) -> "Uri":
```

Comment:
I think we should enforce that message is a protobuf message, and get the type information from the protobuf, not python module, so that it's the same in c++ and python

Reply:
sure, open to discuss implementation details for this